pf/internal/defaults/defaults.go

```diff
@@ -0,0 +1,293 @@
+	}
+	return res
+}
+
+// Returns matching object schema for a context determined by the PropertyPath, if any.
+func (du *defaultsTransform) lookupSchemaByContext(
+	path resource.PropertyPath,
+) (shim.SchemaMap, map[string]*tfbridge.SchemaInfo, bool) {
+	if len(path) == 0 {
+		return du.topSchemaMap, du.topFieldInfos, true
+	}
+
+	schemaPath := tfbridge.PropertyPathToSchemaPath(path, du.topSchemaMap, du.topFieldInfos)
+	if schemaPath == nil {
+		return nil, nil, false
+	}
+
+	schema, info, err := tfbridge.LookupSchemas(schemaPath, du.topSchemaMap, du.topFieldInfos)
+	if err != nil {
+		return nil, nil, false
+	}
+
+	encodedObjectSchema, ok := schema.Elem().(shim.Resource)
+	if !ok {
+		return nil, nil, false
+	}
+
+	objectSchema := encodedObjectSchema.Schema()
+
+	var fields map[string]*tfbridge.SchemaInfo
+	if info == nil || info.Fields == nil {
+		fields = map[string]*tfbridge.SchemaInfo{}
+	} else {
+		fields = info.Fields
+	}
+
+	return objectSchema, fields, true
+}
+
+// Extends PropertyMap with Pulumi-specified default values.
+func (du *defaultsTransform) extendPropertyMapWithDefaults(
+	ctx context.Context,
+	path resource.PropertyPath,
+	props resource.PropertyMap,
+) (resource.PropertyMap, error) {
+	schemaMap, infos, ok := du.lookupSchemaByContext(path)
+	if !ok {
+		return props, nil
+	}
+	res := props.Copy()
+
+	// Can iterate over SchemaInfo.Fields instead of iterating over every field in the schema, since the algorithm
+	// is only interested in properties where there is a SchemaInfo.Field specifying a default.
+	for key, fld := range infos {
+		if fld == nil || fld.Default == nil {
+			continue
+		}
+		fieldSchema, knownField := schemaMap.GetOk(key)
+		if !knownField {
+			continue
+		}
+
+		pulumiName := tfbridge.TerraformToPulumiNameV2(key, schemaMap, infos)
+		pk := resource.PropertyKey(pulumiName)
+
+		if _, setAlready := res[pk]; setAlready {
+			continue
+		}
+
+		// using default value for empty property
+		pv, gotDefault, err := getDefaultValue(ctx,
+			du.resourceByPath(path),
+			fieldSchema,
+			fld.Default,
+			du.providerConfig)
+		if err != nil {
+			return nil, fmt.Errorf("when computing a default for property '%s' %w", key, err)
+		}
+		if gotDefault {
+			res[pk] = pv
+		}
+	}
+
+	return res, nil
+}
+
+// This is mostly simply a recursion pattern on PropertyValue, can be abstracted out.
+func (du *defaultsTransform) withDefaults(
+	ctx context.Context,
+	path resource.PropertyPath,
+	value resource.PropertyValue,
+) resource.PropertyValue {
+	tr := func(path resource.PropertyPath, value resource.PropertyValue) (resource.PropertyValue, error) {
+		if !value.IsObject() {
+			return value, nil
+		}
+		pm := value.ObjectValue()
+		// After recurring on the elements, try to apply defaults here.
+		if extended, err := du.extendPropertyMapWithDefaults(ctx, path, pm); err != nil {
+			tflog.Trace(ctx, "Ignoring an error in extendPropertyMapWithDefaults",
+				map[string]any{"err": err.Error()})
```

Comment:
Does pkg/tfbrige log errors and continue on when they fail to parse values?

Reply:
It fails e.g. `strconv.ParseBool: parsing "FOOBAR": invalid syntax` we can match that here.

Reply:
I've tinkered a bit and decided to flush these down as `tflog.Error` - this should surface nicely in the engine, without stopping the program as in a catastrophic error; nice advantage is that if you have many of these, the user can be exposed to several error messages at once, instead of debugging one by one.